[RFC] Add JSON support

ZyX-I opened the pull request:
- [x] jsonencode()
- [x] jsondecode()
- [x] v:true, v:null and v:false
- [x] jsondecode() returning special dictionary for strings with NUL bytes
- [x] jsondecode() returning special dictionary for mappings with duplicate key
- [x] jsondecode() returning special dictionary for mappings with key with NUL byte

ZyX-I committed:
eval: Split out typval_T dumping functions to nvim/encode.c

ZyX-I committed:
functests(msgpack): Fix test names

ZyX-I committed:
functests(msgpack): Fix location of one of the tests

ZyX-I committed:
eval: Add jsonencode() function

Ref #3471

oni-link commented on src/nvim/encode.c:
```diff
@@ -0,0 +1,1176 @@
+  while (kv_size(mpstack)) { \
+    MPConvStackVal *cur_mpsv = &kv_A(mpstack, kv_size(mpstack) - 1); \
+    typval_T *cur_tv = NULL; \
+    switch (cur_mpsv->type) { \
```
Can you add here a `default` case or an `assert(cur_tv)` after the switch? Otherwise the static analyzer will complain again that we could call `name##_convert_one_value(firstargname, &mpstack, cur_tv, copyID, objname))` with a `NULL` argument.

ZyX-I committed:
encode: Add a hint for static analyzer that cur_tv is not NULL

ZyX-I committed:
eval: Add special variable type

ZyX-I committed:
eval: Add special variables v:false, v:null, v:none

ZyX-I committed:
runtime/msgpack: Add support for special values

ZyX-I committed:
functests/msgpack: Test dumping special nil and bool dicts

ZyX-I commented:
Note that I changed typval_T contents: now it contains two enums in place of two `char`s, also where there is `switch(…v_type)` I removed `default` cases in order to get compiler notifications about missing enum values.

ZyX-I committed:
eval: Add jsondecode() function

ZyX-I committed:
encode: Do not use default case for v_type

ZyX-I committed:
cmake: Refactor code that creates directories and lists C files

ZyX-I committed:
eval: Move encode.c to eval/encode.c

ZyX-I committed:
vim: Move *MSG* macros to message.h

Note: OUT* macros were removed because they are no longer used anywhere.

ZyX-I committed:
eval: Get rid of VV_LEN constant

ZyX-I committed:
eval: Move some decoding functions to eval/decode.c

ZyX-I committed:
eval/decode: Replace INIT_SPECIAL_DICT macros with inline function

ZyX-I committed:
eval/decode: Make msgpackparse() function use new v: vars

ZyX-I committed:
eval/decode: Fail on control and invalid unicode characters

ZyX-I committed:
eval/decode: Fix surrogate pairs processing

ZyX-I committed:
doc: Update documentation regarding converting from/to &encoding

ZyX-I committed:
eval/*: Fix some linter errors

ZyX-I commented:
@neovim/core I have a question: what to do with new v: values and Object? There are two issues:

1. Currently AFAIK each different typval_T produces different Objects, but v:none does not fit here (v:null can and is currently mapped to nil object). I think v:none should be simply removed since parser was not going to produce it in any case.
2. Worse problem is what to do with APIs like `set_var`: currently passing nil unsets the variable, (BTW, I cannot do this from lua), but after this PR I would expect passing `nil` to set variable to `v:null`.

ZyX-I committed:
eval/decode: Parse strings with NUL to special dictionaries

bfredl commented:
`nil` in lua seems to mean a nonextisting value, rather than an explicit null value, for instance `#{nil, nil, nil} == 0` and `t["key"] = nil` means deleting the key. So I think the present `set_var` behavior is the most expected from a _lua_ point of view (but not necessarily the best from a _nvim_ point of view).

ZyX-I commented:
@bfredl Currently lua is only used for tests and building something, and there exists neovim/lua-client#24 problem which makes your reply have no sense so far. NIL in msgpack does not have such semantics (it is up to higher level code that uses msgpack to decide what semantics NIL has, otherwise it is a regular value).

bfredl commented:
oh, you're right, there is both the lua <=> msgpack and mskpack <=> vimL conversions going on here... msgpack NIL is definitely like None/v:null/etc. But it still holds that lua `nil` is pretty much like `v:none` (no matter how much of completely pointless the later is, I completely I agree with that)

ZyX-I commented:
@bfredl Master conversion code converts NIL to 0, so it does not have this semantics in Neovim as well.

bfredl commented:
Yes, right now, of course, but it would make sense to change it to `v:null` when we have it.

ZyX-I committed:
eval/decode: Add support for special maps

Special dictionaries representing map are created when encountering duplicate
key or when key is empty or contains NUL.

Also checks that values are separated by a comma/colon properly.

ZyX-I committed:
eval: Remove get_vim_var_tv function

ZyX-I commented:
PR is finished, except for the question regarding what to do with vim_set_var({name}, NIL).

ZyX-I committed:
eval: Remove v:none

To get v:none back just rever this commit. This will not make json*() functions
compatible with Vim though.

ZyX-I committed:
*: Fix linter errors

ZyX-I committed:
eval/decode: Fix vim_str2nr invocation

ZyX-I committed:
eval: Change dv_lock type to VarLockStatus

ZyX-I committed:
*: Fix gcc warnings

oni-link commented on src/nvim/api/private/helpers.c:
```diff
@@ -397,13 +397,13 @@ bool object_to_vim(Object obj, typval_T *tv, Error *err)
-      tv->v_type = VAR_NUMBER;
-      tv->vval.v_number = 0;
+      tv->v_type = VAR_SPECIAL;
+      tv->vval.v_number = kSpecialVarNull;
```
Is `tv->vval.v_number` the right place to save a special value? In new line 655, special values are read from `...->vval.v_speical`.

ZyX-I committed:
eval: Make assert_true and assert_false accept v:true and v:false

ZyX-I committed:
eval: Port parts of 7.4.1267 that are not already present

ZyX-I committed:
shada: Make sure that NIL and EXT values can also be parsed back

Note: currently they are both *dumped*, but parsing them produces an error. This
is inappropriate: variables should either be skipped with error message when
dumping or should be read back properly.

It also appears that I did not have test for “has wrong variable value type”
error, so nothing got removed from errors_spec.

ZyX-I committed:
*: Fix memory leaks found by clang sanitizer

ZyX-I committed:
eval/encode: Fix non-utf-8 &encoding handling, add tests

ZyX-I committed:
version: Record that vim-7.4.1271 is kinda of there

ZyX-I committed:
eval: Fix QuickBuild failures

Compiler used by one VM in QuickBuild has found a number of false positives.
Everything is fine on travis.

List of failures:

From [QuickBuild][1], build [7429][2]:

    14:38:19,945 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c: In function ‘assert_bool’:
    14:38:19,945 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c:7551:40: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
    14:38:20,058 WARN  - cc1: all warnings being treated as errors

. This is not making much sense (7551:40 is `!=` in `{SpecialVarValue} !=
({bool}?{SpecialVarValue}:{SpecialVarValue})`), but this error is present.

---

Also fail from [build][3] [4930][4]:

    15:47:00,853 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval/encode.c: In function ‘encode_read_from_list’:
    15:47:00,853 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval/encode.c:258:30: error: conversion to ‘char’ from ‘int’ may alter its value [-Werror=conversion]

, pointing to `:` in `{char} = ({char} == {const} ? {const} : {char})` where
`{const}` is character constant like `'\n'`. I have no idea where exactly it saw
conversion, so simply casted everything to (char).

---

[Build][5] error:

    08:32:03,472 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c: In function ‘tv_equal’:
    08:32:03,472 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c:5077:1: error: control reaches end of non-void function [-Werror=return-type]

---

Build [4949][7]:

    11:28:00,578 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c: In function ‘f_type’:
    11:28:00,578 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c:16085:24: error: ‘n’ may be used uninitialized in this function [-Werror=uninitialized]
    11:28:00,581 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c: In function ‘f_empty’:
    11:28:00,581 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c:8505:24: error: ‘n’ may be used uninitialized in this function [-Werror=uninitialized]

[1]: http://neovim-qb.szakmeister.net/wicket/page?5-1.ILinkListener-content-buildTab-panel-masterStep-body-children-0-step-body-children-2-body-children-3-step-body-children-0-step-body-children-0-step-head-logLink
[2]: http://neovim-qb.szakmeister.net/build/4929
[3]: http://neovim-qb.szakmeister.net/build/4930
[4]: http://neovim-qb.szakmeister.net/wicket/page?1-1.ILinkListener-content-buildTab-panel-masterStep-body-children-0-step-body-children-1-body-children-3-step-body-children-0-step-body-children-0-step-head-logLink
[5]: http://neovim-qb.szakmeister.net/build/4948/step_status
[7]: http://neovim-qb.szakmeister.net/build/4949

justinmk commented:
> I think v:none should be simply removed since parser was not going to produce it in any case.

Removed, as in Neovim does not support it at all? Fine with me. 

> what to do with APIs like set_var: currently passing nil unsets the variable

Before now setting a VimL var to "nil" was not valid, but now that it is, the API behavior should be changed. `:unlet` and `remove()` exist, and if we want first-class support (as opposed to "eval") for these things in the API then we will eventually add it for all functions, not special cases.

> Master conversion code converts NIL to 0

Probably anywhere that NIL is sent it should send `v:null`. If `v:null` is not valid for the variable type then it should raise an error, not degrade to 0. We should probably save that for another PR though.

ZyX-I commented:
@justinmk I have already changed code to convert nil to v:null. Question was what to do with existing behaviour right now because I think that clients rely on it. I also do not know what besides vim_set_var has similar special-casing.

Also CC @tarruda, I guess he knows better.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1260 @@
+    const char *line_start = line_end;
+    line_end = xmemscan(line_start, NL, (size_t) (end - line_start));
+    if (line_end == line_start) {
+      list_append_allocated_string(list, NULL);
```
`li != NULL` seems to indicate that the string in the last list item must be completed. If the
first byte in `buf` is a `NL`,  this is not necessary. `li` should be set to `NULL` here.

ZyX-I replied:
It is more complicated then this. I am now writing unit tests, but it seems that at least the following change is needed:

```Patch
diff --git a/src/nvim/eval/encode.c b/src/nvim/eval/encode.c
index 2fd1c9d..fb3c237 100644
--- a/src/nvim/eval/encode.c
+++ b/src/nvim/eval/encode.c
@@ -80,7 +80,11 @@ int encode_list_write(void *data, const char *buf, size_t len)
     const char *line_start = line_end;
     line_end = xmemscan(line_start, NL, (size_t) (end - line_start));
     if (line_end == line_start) {
-      list_append_allocated_string(list, NULL);
+      if (li) {
+        li = NULL;
+      } else {
+        list_append_allocated_string(list, NULL);
+      }
     } else {
       const size_t line_length = (size_t) (line_end - line_start);
       char *str;
```

ZyX-I replied:
Should be fixed now. I confirmed than new tests fail without new change.

oni-link replied:
Should `encode_list_write(data,"\n\n\n",3);` and 
`encode_list_write(data,"\n",1);encode_list_write(data,"\n",1);encode_list_write(data,"\n",1);` result in the same list?

ZyX-I replied:
Yes, my patch is making things worse. Though final variant (pushed) is simpler: I just need the exact same construct as in the end of `else` clause (to `if (line_end == line_start)`).

ZyX-I committed:
eval/encode: Fix writing strings starting with NL to list

Error [found][1] by oni-link.

[1]: https://github.com/neovim/neovim/pull/4131/files#r52239384

tarruda commented:
> Question was what to do with existing behaviour right now because I think that clients rely on it. I also do not know what besides vim_set_var has similar special-casing.

If this change is required to support json semantics better, then I see no other choice. But we must add `{vim,buffer,tabpage,window}_del_var` to support deleting variables.

To avoid compatibility issues with vim-python plugins, we can add a [session hook](https://github.com/neovim/python-client/blob/dea5427259ecdbbccd058a2ca0b6c7267138816c/neovim/api/common.py#L123-L140) that intercepts `vim_set_var` with `None` as the value argument and changes the call to `vim_del_var`. There's a working [example](https://github.com/neovim/neovim/blob/9b9695ab8e00eda21ed77dcc9cfb2800f35b268b/runtime/autoload/provider/script_host.py#L166-L180) that emulates the eval behavior from vim-python, but the SessionHook infrastructure must be adapted to allow changing the method name(and not only the arguments).

[RemoteMap](https://github.com/neovim/python-client/blob/dea5427259ecdbbccd058a2ca0b6c7267138816c/neovim/api/common.py#L23) must also be updated so that `del vim.vars` will invoke the new function.

bfredl commented:
Why? legacy vim-python plugins won't call vim_set_var directly, only use the RemoteMap interfaces, so updating the later should be enough.

tarruda commented:
> Why? legacy vim-python plugins won't call vim_set_var directly, only use the RemoteMap interfaces, so updating the later should be enough.

True, but some vim-python plugins might rely on `vim.vvars['key'] = None` deleting keys(AFAIK, I've copied this behavior from vim-python)

bfredl commented:
Yes, but I meant this check could be done in `RemoteMap` that defines the user-level interface rather than messing with our own API.

tarruda commented:
> Yes, but I meant this check could be done in RemoteMap that defines the user-level interface rather than messing with our own API.

Your'e right, this is simpler than messing with session hooks. If a python remote plugin needs to set a variable to `v:null` they can do so by calling `vim_set_var` directly.

bfredl commented:
actually `:py vim.vars['xx'] = None` is an error in recent vim. Or do you mean something else?

tarruda commented:
> actually :py vim.vars['xx'] = None is an error in recent vim. Or do you mean something else?

This is what I mean. @ZyX-I, has this behavior changed recently or I simply didn't emulate vim-python correctly?

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+  const char *line_end = buf;
+  if (list->lv_last == NULL) {
+    list_append_string(list, NULL, 0);
+  }
```
This `if` block could be removed. If the list is empty, the loop code will add the first element.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+        if (str[i] == NUL) {
+          str[i] = NL;
+        }
+      }
```
This `for` loop could be replaced with `memchrsub(str, NUL, NL, line_length);`.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+        const size_t li_len = (li->li_tv.vval.v_string == NULL
+                               ? 0
+                               : STRLEN(li->li_tv.vval.v_string));
+        li->li_tv.vval.v_string = xrealloc(li->li_tv.vval.v_string,
```
The lines above could be made a bit more readable like this:
```c
str = (char *)li->li_tv.vval.v_string;
const size_t li_len = (str == NULL ? 0 : STRLEN(str));
li->li_tv.vval.v_string = xrealloc(str, li_len + line_length + 1);
```

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+          if (buf_[i] == '\'') { \
+            num_quotes++; \
+          } \
+        } \
```
Could be replaced with `memcnt(buf_,'\'',len_);`.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+        ga_append(gap, '\''); \
+        for (size_t i = 0; i < len_; i++) { \
+          if (buf_[i] == '\'') { \
+            num_quotes++; \
```
Can be removed.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+                    "UTF-8 character"), utf_buf);
+            return FAIL;
+          } else if ((0xD800 <= ch && ch <= 0xDB7F)
+                     || (0xDC00 <= ch && ch <= 0xDFFF)) {
```
Why are some surrogate code points excluded? Range should be from `0xD800` to `0xDFFF`.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+          if (ch > 0x7F && shift == 1) {
+            EMSG2(_("E474: String \"%s\" contains byte that does not start any "
+                    "UTF-8 character"), utf_buf);
+            return FAIL;
```
Missing `xfree(tofree);`?

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1258 @@
+                     || (0xDC00 <= ch && ch <= 0xDFFF)) {
+            EMSG2(_("E474: UTF-8 string contains code point which belongs "
+                    "to surrogate pairs: %s"), utf_buf + i);
+            return FAIL;
```
Missing `xfree(tofree);`?

ZyX-I replied:
> Missing xfree(tofree);?

Not actually: these conditions should not happen after string_convert with .vc_fail=true. Though I will add them just in case.

ZyX-I committed:
eval/encode: Fix invalid UTF-8 strings handling:

1. Do not allow reading past buffer end when creating error messages.
2. Fix surrogate pairs range, avoid magic constants.

ZyX-I committed:
eval/encode: Reduce length of encode_list_write

Changes suggested by oni-link.

ZyX-I committed:
eval/encode: Free memory just in case

After string_convert() with .vc_fail=true these blocks should never be entered 
because they indicate invalid unicode.

ZyX-I committed:
eval/encode: Remove unneeded variable, add missing include

ZyX-I commented:
@tarruda There is another problem: `*_set_var` methods not only accept NILs, they *return* NILs to indicate that there was no previous value. This way variable set to NIL cannot be distinguished with unset variable.

ZyX-I committed:
api: Replace set_var(name, NIL) with del_var(name)

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1251 @@
+        return false;
+      }
+      len++;
+      if (li->li_tv.vval.v_string != 0) {
```
`0` -> `NULL`

justinmk commented:
> another problem: `*_set_var` methods not only accept NILs, they *return* NILs to indicate that there was no previous value. This way variable set to NIL cannot be distinguished with unset variable.

`set_var` should return an error if the variable is undefined, just like VimL.

Instead of `*_del_var`, `*_set_var`, `*_get_var` why don't we expose `dict_` functions in the API that can operate on any dict? And for the specific case of "del", if `vim_eval("unlet b:foo")` is not ideal we could add `vim_unlet`.

ZyX-I commented:
@justinmk VimL does *not* return error when you do `let undefined_var = 1`, it creates `undefined_var` (I *did not* mean `meths.set_var('bar', nil)`). This API function is doing the following: it does the equivalent of `let` and returns old value of the variable. I.e. `local foo = meths.set_var('bar', 42)` is like `local foo = meths.get_var('bar'); meths.set_var('bar', 42)`, but with two differences: does not error out when `bar` does not exist and it is not possible to change variable value between `get_var` and `set_var`.

justinmk commented:
I was thinking of `get_var` (I'm rushing too much to get a word in...). Also after reviewing `api/` I think see that my suggestion of `dict_` functions is not that great.

ZyX-I commented:
@justinmk And about `dict_` functions: this was already discussed when `vim.bindeval` implementation was discussed. I do not like `vim_unlet` because this function automatically means invoking `evalN` functions for parsing, or will be a different thing from `:unlet`.

ZyX-I commented:
`get_var` errors out.

ZyX-I commented:
@justinmk Also note that `buffer_del_var` allows deleting variable from any buffer. `:unlet` does not. `vim_unlet` may accept an `Object` that is either tabpage, window or buffer, but this is not consistent with other API AFAIK.

justinmk commented:
> vim_unlet may accept an Object that is either tabpage, window or buffer, but this is not consistent with other API AFAIK.

That's why I am not feeling so great about the `dict_` functions now. But maybe I will explore it with a PR.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1251 @@
+  while (p < buf_end) {
+    for (size_t i = state->offset; i < state->li_length && p < buf_end; i++) {
+      const char ch = (char) state->li->li_tv.vval.v_string[state->offset++];
+      *p++ = (char) ((char) ch == (char) NL ? (char) NUL : (char) ch);
```
Do we need any of these casts?

ZyX-I replied:
Check the associated commit. We do or quickbuild failes, but I do not know which of them are needed exactly.

wsdjeg commented:
should we change the function name to `*json_decode()`,just like vim

ZyX-I commented:
@wsdjeg Renamed.

ZyX-I committed:
eval: Rename json* functions to json_*

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,941 @@
+        .vval = { .v_list = list },
+      }));
+      if (encode_list_write((void *) list, mobj.via.str.ptr, mobj.via.str.size)
+          == -1) {
```
`encode_list_write()` always returns `0`.

ZyX-I replied:
Interface (derived from msgpack) assumes that it *may* return -1. So I am not changing this.

wsdjeg commented:
thanks

ZyX-I commented:
I am marking this as RDY.

@tarruda @justinmk  Regarding return value of `set_var`: I can say that `dict_set_value` may return `ArrayOf(Object, 2)` (basically, a pair `(Boolean, Object)`, but I cannot express this currently): first value says whether there was any previous value, second value contains that previous value. Other then this point PR is ready.

ZyX-I committed:
api/documentation: Add a warning that nil may mean v:null

ZyX-I committed:
encode: Fail to dump NaN and infinity

Thanks to vim/vim#654

ZyX-I committed:
encode: Handle incomplete surrogates like `\uSURR\uOTHR` properly

ZyX-I committed:
eval: Do not break when VimVarIndex and vimvars order mismatches

Also makes sure that compiler will error out when new name is longer then 
vv_filler.

ZyX-I committed:
eval: Remove assert expression

GCC on travis thinks that
1. It is not constant.
2. Left-hand operand of comma has no effect (-Werror=unused-variable).

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+  dict_T *const dict = dict_alloc();
+  dictitem_T *const type_di = dictitem_alloc((char_u *) "_TYPE");
+  type_di->di_tv.v_type = VAR_LIST;
+  type_di->di_tv.v_lock = 0;
```
`VAR_UNLOCKED` instead of `0`.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+
+/// Helper function used for working with stack vectors used by JSON decoder
+///
+/// @param[in]  obj  New object.
```
Missing `out`:
`obj` is a copy, but it contains pointers that could be freed (`clear_tv(&obj.val)`).

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+        }
+        ContainerStackItem last_container = kv_last(container_stack);
+        if (*p == '}' && last_container.container.v_type != VAR_DICT) {
+          EMSG2(_("E474: Closing list with figure brace: %s"), p);
```
figure brace == curly brace ?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+        continue;
+      }
+      case 'n': {
+        if (strncmp(p + 1, "ull", 3) != 0) {
```
Is `p` NUL terminated? Otherwise the following byte positions could already be outside of the buffer.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+        break;
+      }
+      case 't': {
+        if (strncmp(p + 1, "rue", 3) != 0) {
```
`p` NUL terminated?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+        break;
+      }
+      case 'f': {
+        if (strncmp(p + 1, "alse", 4) != 0) {
```
`p` NUL terminated?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+            // and thus cannot possibly be equal to *p. But utf_ptr2char({0xFF,
+            // 0}) will return 0xFF, even though 0xFF cannot start any UTF-8
+            // code point at all.
+            if (ch >= 0x80 && p_byte == ch) {
```
That is not always right: `U+00C3` has byte encoding `0xC3 0x83`, so with `p_byte == ch` we could detect an UTF-8 character as an encoding error.

Also see here:
https://github.com/neovim/neovim/blob/master/src/nvim/mbyte.c#L1199-L1204

oni-link commented:
With `:echo json_decode("")` I get the following assertion:
```bash
nvim: /home/oni-link/git/neovim/src/nvim/eval/decode.c:723: int json_decode_string(const char *const, const size_t, typval_T *const): Assertion `((stack).size) == 1' failed.
```

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+            p += ch_len;
+          }
+        }
+        if (*p != '"') {
```
Do we need to check for `p < e` here before dereferencing the pointer?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+          if (*p == '\\') {
+            p++;
+            if (p == e) {
+              EMSG2(_("E474: Unfinished escape sequence: %s"), buf);
```
Is `buf` NUL terminated?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+                  goto json_decode_string_fail;
+                }
+                // One UTF-8 character below U+10000 can take up to 3 bytes,
+                // above up to 6, but they are encoded using two \u escapes.
```
UTF-8 only uses up to 4 bytes to encode a valid UTF-8 code point (since 2003?).

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+                  hasnul = true;
+                }
+                if (SURROGATE_HI_START <= ch && ch <= SURROGATE_HI_END) {
+                  fst_in_pair = (int) ch;
```
Don't we lose the first character of "\uD800\uD800" (both characters in the high surrogate range)?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+        }
+        if (p < e && *p == '.') {
+          p++;
+          fracs = p;
```
Is "-.1" a valid JSON number?

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+        } else if (p == exps) {
+          EMSG2(_("E474: Missing exponent: %s"), s);
+          goto json_decode_string_fail;
+        }
```
What about this numbers, are they valid JSON numbers?
`10e+1`, `10.e+1`

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,945 @@
+      }
+    }
+  }
+  if (kv_size(stack) > 1 || kv_size(container_stack)) {
```
Could be rewritten to be less "jumpy":

```c
  if (kv_size(stack) == 1 && kv_size(container_stack) == 0) {
    *rettv = kv_pop(stack).val;
    goto json_decode_string_ret;
  }
  EMSG2(_("E474: Unexpected end of input: %s"), buf);

json_decode_string_fail:
  ret = FAIL;
  while (kv_size(stack)) {
    clear_tv(&(kv_pop(stack).val));
  }

json_decode_string_ret:
  kv_destroy(stack);
  kv_destroy(container_stack);
  return ret;
}
```

ZyX-I committed:
eval/decode: Also use VAR_UNLOCKED in old code

ZyX-I committed:
eval/decode: Record that `obj` may be freed

ZyX-I committed:
eval/decode: Rename brackets in error messages

U+007D is officially RIGHT CURLY BRACKET.
U+005D is officially RIGHT SQUARE BRACKET.

ZyX-I committed:
eval/decode: Avoid overflow when parsing incomplete null/true/false

Note: second test does not crash or produce asan errors, even though it should.

ZyX-I committed:
functests: Make sure that json functions are tested with C messages

ZyX-I committed:
eval/decode: Make sure that U+00C3 is parsed correctly

ZyX-I committed:
eval/decode: Make sure that blank input does not crash Neovim

ZyX-I committed:
eval/decode: Make sure that parsing strings does not overflow

ZyX-I committed:
eval/decode: Make sure that error messages do not cause overflow

ZyX-I committed:
eval/decode: Do not loose high surrogates followed by high surrogates

ZyX-I committed:
eval/decode: Reject more numbers, accept 1e5

ZyX-I committed:
eval/decode: Rewrite json_decode_string end as suggested by oni-link

ZyX-I committed:
message: Remove useless emsg* functions, leaving only emsgf

ZyX-I committed:
documentation: Update type() documentation

ZyX-I committed:
*: Fix linter errors

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1094 @@
+    p++;
+  }
+  ints = p;
+  while (p < e && ascii_isdigit(*p)) {
```
Only one leading `0` is allowed.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1094 @@
+  if (p == ints) {
+    emsgf(_("E474: Missing number after minus sign: %.*s"), LENP(s, e));
+    goto parse_json_number_fail;
+  } else if (p == fracs || exps == fracs + 1) {
```
`1.e+` errors out as missing exponent instead of missing number after dot.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1094 @@
+    vim_str2nr((char_u *) s, NULL, &num_len, 0, &nr, NULL, (int) (p - s));
+    if ((int) exp_num_len != num_len) {
+      emsgf(_("E685: internal error: while converting number \"%.*s\" "
+              "to float vim_str2nr consumed %i bytes in place of %zu"),
```
Does not convert to float.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1094 @@
+/// @param[in]  buf_len  Length of the buffer.
+/// @param[in,out]  pp  Pointer to the start of the number. Must point to
+///                     a digit or a minus sign. Is advanced to the last
+///                     character of the number.
```
[out]: In case of an error this does not point to a number (`1.`)

ZyX-I replied:
09.03.2016, 01:34, "oni-link" <notifications@github.com>:In src/nvim/eval/decode.c:> +parse_json_string_ret:
> +  *pp = p;
> +  return ret;
> +}
> +
> +#undef POP
> +
> +/// Parse JSON number: both floating-point and integer
> +///
> +/// Number format: `-?\d+(?:.\d+)?(?:[eE][+-]?\d+)?`.
> +///
> +/// @param[in]  buf  Buffer being converted.
> +/// @param[in]  buf_len  Length of the buffer.
> +/// @param[in,out]  pp  Pointer to the start of the number. Must point to
> +///                     a digit or a minus sign. Is advanced to the last
> +///                     character of the number.
[out]: In case of an error this does not point to a number (1.) ? —Reply to this email directly or view it on GitHub. 

ZyX-I replied:
In case of error nobody cares what it points to. Though there is a “restart” case when it will point to the first character of the dictionary.

ZyX-I committed:
eval/decode: Reject even more numbers

Rejects leading zeroes and numbers like 1.e+5 (decimal dot with missing number 
with signed exponent).

ZyX-I committed:
eval/decode: Fix typo in internal error message

ZyX-I committed:
eval/decode: Clarify meaning of some pointer arguments

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1106 @@
+  while (p < e && ascii_isdigit(*p)) {
+    p++;
+  }
+  if (p != ints + 1 && *ints == '0') {
```
`ints == e` if nothing comes after `-`.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1106 @@
+    emsgf(_("E474: Leading zeroes are not allowed: %.*s"), LENP(s, e));
+    goto parse_json_number_fail;
+  }
+  if (p < e && p != ints && (*p == '.' || *p == 'e' || *p == 'E')) {
```
This test `(*p == '.' || *p == 'e' || *p == 'E')` could be removed.

ZyX-I committed:
eval/decode: Do not overflow when parsing `-`

Also makes if’s less nested.

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1114 @@
+{
+  const char *p = buf;
+  const char *const e = buf + buf_len;
+  while (p < e && (*p == ' ' || *p == '\t' || *p == '\n')) {
```
Looks like `CAR` (U+000D) could also be used as whitespace.

ZyX-I committed:
eval/decode: Accept `\r` as space character

oni-link commented on src/nvim/eval/decode.c:
```diff
@@ -0,0 +1,1116 @@
+          goto json_decode_string_fail;
+        } else if (last_container.stack_index != kv_size(stack) - 1) {
+          assert(last_container.stack_index < kv_size(stack) - 1);
+          emsgf(_("E474: Expected value: %.*s"), LENP(p, e));
```
Don't we expect a `':'` in this case?

ZyX-I replied:
We expect both `:` and value. Value is more significant: e.g. in Python `{"a", "b"}` is a valid set literal.

oni-link replied:
Normally the parser tells us exactly what it wanted to see next. Could also be helpful when searching for the error: `'{"1":1,"2" 2}'`

ZyX-I replied:
@oni-link For `'{"1": 1, "2" 2}'` it errors out in different place. And with the current code one does not need to search for error, unlike Bram’s my code tells what error is and where it is.

ZyX-I committed:
documentation,functests: State that UTF-8-only support is intentional

oni-link commented on src/nvim/api/tabpage.c:
```diff
@@ -69,7 +72,24 @@ Object tabpage_set_var(Tabpage tabpage, String name, Object value, Error *err)
+/// @param tabpage handle
+/// @param name The variable name
+/// @param[out] err Details of an error that may have occurred
+/// @return The tab page handle
```
`/// @return the old value, if any`

ZyX-I committed:
api: Add warnings about the returned value to \*del_var functions

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1290 @@
+  [CAR] = "\\r",
+  ['"'] = "\\\"",
+  ['\\'] = "\\\\",
+};
```
Entry for `[FF] ` is missing.

oni-link replied:
`escapes`-entry with index `FF` is used in `convert_to_json_string()`:
```c
      switch (ch) {
        case BS:
        case TAB:
        case NL:
        case FF:
        case CAR:
        case '"':
        case '\\': {
          ga_concat_len(gap, escapes[ch], 2);
          break;
        }
        default: {
...
```

oni-link commented on src/nvim/eval_defs.h:
```diff
@@ -16,39 +16,52 @@ typedef double float_T;
+  VAR_LIST,         ///< List, .v_list is used.
+  VAR_DICT,         ///< Dictionary, .v_dict is used.
+  VAR_FLOAT,        ///< Floating-point value, .v_float is used.
+  VAR_SPECIAL,      ///< Special value (true, false, null, none), .v_special
```
`none` does not seem to be a value of `VAR_SPECIAL`.

ZyX-I replied:
20.03.2016, 16:22, "oni-link" <notifications@github.com>:In src/nvim/eval_defs.h:> +typedef enum {
> +  VAR_UNLOCKED = 0,  ///< Not locked.
> +  VAR_LOCKED,        ///< User lock, can be unlocked.
> +  VAR_FIXED,         ///< Locked forever.
> +} VarLockStatus;
> +
> +/// VimL variable types, for use in typval_T.v_type
> +typedef enum {
> +  VAR_UNKNOWN = 0,  ///< Unknown (unspecified) value.
> +  VAR_NUMBER,       ///< Number, .v_number is used.
> +  VAR_STRING,       ///< String, .v_string is used.
> +  VAR_FUNC,         ///< Function referene, .v_string is used for function name.
> +  VAR_LIST,         ///< List, .v_list is used.
> +  VAR_DICT,         ///< Dictionary, .v_dict is used.
> +  VAR_FLOAT,        ///< Floating-point value, .v_float is used.
> +  VAR_SPECIAL,      ///< Special value (true, false, null, none), .v_special
none does not seem to be a value of VAR_SPECIAL. Fixed as a part of `eval: Remove v:none`. —You are receiving this because you were mentioned.Reply to this email directly or view it on GitHub 

oni-link commented on src/nvim/eval.c:
```diff
@@ -7586,16 +7586,20 @@ static void f_assert_equal(typval_T *argvars, typval_T *rettv)
+      && (argvars[0].v_type != VAR_SPECIAL
+          || argvars[0].vval.v_special != (is_true
+                                           ?kSpecialVarTrue
+                                           :kSpecialVarFalse))) {
```
Help needs to be updated for using special values in `assert_{true,false}()`:

The help for `assert_false()` says `A value is false when it is zero. When "{actual}" is not a number the assert fails.`. This could mean a fail for all special values or, if the described conversion in the help for special values is used, that **`v:null`** and `v:false` should not fail, because they are converted to `0`.

oni-link commented on src/nvim/eval.c:
```diff
@@ -8473,6 +8473,9 @@ static void f_empty(typval_T *argvars, typval_T *rettv)
     break;
+  case VAR_SPECIAL:
+    n = argvars[0].vval.v_special != kSpecialVarTrue;
+    break;
```
Does this need a help update for special vars or is the conversion to numbers being used?

ZyX-I committed:
documentation: Update assert_{false,true}() and empty() documentation

Also removes a note regarding the performance of `empty(long_list)` vs
`len(long_list) == 0` because this has nothing to do with the actual state:
first checks that list is not NULL and pointer to its first element is also not
NULL, second gets length by comparing list with NULL and falls back to
`tv->vval.v_list->lv_len` if not. `len(long_list)` *may* still be *slightly*
slower, but the slow down has nothing to do with the length of the list, is
hardly noticeable and depends on how good compiler is at inlining and what
exactly have author of the plugin written (I mean `len(long_list) == 0` vs
`empty(long_list)` vs `!len(long_list)`).

oni-link commented on src/nvim/eval.c:
```diff
@@ -11558,6 +11570,529 @@ static void f_join(typval_T *argvars, typval_T *rettv)
+  char *tofree = NULL;
+  size_t len;
+  if (argvars[0].v_type == VAR_LIST) {
+    if (!encode_vim_list_to_buf(argvars[0].vval.v_list, &len, &s)) {
```
For input `[]` and `['']`, `json_decode_string()` returns `0` (valid JSON?).
If input is `['','']` it crashes with

```c
==13267==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200016ffb1 at pc 0x0000004726fe bp 0x7fff35630590 sp 0x7fff3562fd40
READ of size 2 at 0x60200016ffb1 thread T0
    #0 0x4726fd in __interceptor_strlen /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:213
    #1 0x1642a6c in vim_vsnprintf /home/oni-link/git/neovim/src/nvim/message.c:3240:25
    #2 0x164bb7d in emsgf /home/oni-link/git/neovim/src/nvim/message.c:574:3
    #3 0xeed7a0 in f_json_decode /home/oni-link/git/neovim/src/nvim/eval.c:11621:5
    #4 0xe01cd3 in call_func /home/oni-link/git/neovim/src/nvim/eval.c:7242:11
    #5 0xe15c74 in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7091:11
    #6 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #7 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #8 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #9 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #10 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #11 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #12 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #13 0xe1599f in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7077:9
    #14 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #15 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #16 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #17 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #18 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #19 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #20 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #21 0xe50e9e in ex_echo /home/oni-link/git/neovim/src/nvim/eval.c:18349:9
    #22 0x135eef0 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2191:5
    #23 0x133d74d in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:601:20
    #24 0x1010a5b in nv_colon /home/oni-link/git/neovim/src/nvim/normal.c:4485:18
    #25 0xff7fb3 in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #26 0x125c920 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:55:26
    #27 0xfac4c0 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #28 0xf77c06 in main /home/oni-link/git/neovim/src/nvim/main.c:538:3
    #29 0x7ff4ab3bd72f in __libc_start_main (/lib64/libc.so.6+0x2072f)
    #30 0x444408 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x444408)

0x60200016ffb1 is located 0 bytes to the right of 1-byte region [0x60200016ffb0,0x60200016ffb1)
allocated by thread T0 here:
    #0 0x4e5088 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:52
    #1 0xaf6124 in try_malloc /home/oni-link/git/neovim/src/nvim/memory.c:59:15
    #2 0xaf62e4 in xmalloc /home/oni-link/git/neovim/src/nvim/memory.c:93:15
    #3 0x177e6c8 in encode_vim_list_to_buf /home/oni-link/git/neovim/src/nvim/eval/encode.c:220:21
    #4 0xeed426 in f_json_decode /home/oni-link/git/neovim/src/nvim/eval.c:11605:10
    #5 0xe01cd3 in call_func /home/oni-link/git/neovim/src/nvim/eval.c:7242:11
    #6 0xe15c74 in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7091:11
    #7 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #8 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #9 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #10 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #11 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #12 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #13 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #14 0xe1599f in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7077:9
    #15 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #16 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #17 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #18 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #19 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #20 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #21 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #22 0xe50e9e in ex_echo /home/oni-link/git/neovim/src/nvim/eval.c:18349:9
    #23 0x135eef0 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2191:5
    #24 0x133d74d in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:601:20
    #25 0x1010a5b in nv_colon /home/oni-link/git/neovim/src/nvim/normal.c:4485:18
    #26 0xff7fb3 in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #27 0x125c920 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:55:26
    #28 0xfac4c0 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #29 0xf77c06 in main /home/oni-link/git/neovim/src/nvim/main.c:538:3

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:213 in __interceptor_strlen

```

oni-link commented on src/nvim/eval.c:
```diff
@@ -17203,6 +17279,12 @@ void set_vim_var_nr(int idx, long val)
 }
 
+/// Set special v: variable to "val"
+void set_vim_var_special(const int idx, const SpecialVarValue val)
```
We could use `VimVarIndex` as type for `idx` for all the `*_vim_var*()` functions.

oni-link commented on src/nvim/eval.c:
```diff
@@ -17548,26 +17551,27 @@ handle_subscript (
-    default:
-      EMSG2(_(e_intern2), "free_tv()");
-      break;
+    switch ((VarType) varp->v_type) {
```
Do we need the cast here?

ZyX-I committed:
eval: Fix overflow in error message in f_json_decode

ZyX-I committed:
eval: Treat [] and [""] as any other empty string

ZyX-I committed:
*: Make set_vim_var_\* functions have proper argument types

ZyX-I committed:
eval/encode: Dump FF character correctly

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -2237,17 +2237,17 @@ assert_equal({expected}, {actual}, [, {msg}])
-		A value is true when it is a non-zeron number.  When {actual}
-		is not a number the assert fails.
+		|v:errors|, like with |assert_equal()|.
+		A value is true when it is a non-zeron number or |v:true|.  
```
Typo: `zeron` -> `zero`

ZyX-I committed:
functests: Replace \xXX escapes with \DDD in lua code

oni-link commented on src/nvim/vim.h:
```diff
@@ -27,6 +27,12 @@ Error: configure did not run properly.Check auto/config.log.
+
+// Can't use "PACKAGE" here, conflicts with a Perl include file.
+#ifndef VIMPACKAGE
+# define VIMPACKAGE     "nvim"
```
Unused macro. Seems we use `PROJECT_NAME` instead.

oni-link commented on src/nvim/fileio.c:
```diff
@@ -4876,35 +4874,39 @@ buf_check_timestamp (
       if (*mesg2 != NUL) {
-        STRCAT(tbuf, "\n");
-        STRCAT(tbuf, mesg2);
+        strncat(tbuf, "\n", tbuf_len);
```
That looks like a wrong use of the length argument of `strncat`.  (`tbuf_len` was previously computed big enough with "some" strings in mind, so it should work none the less.)

We use at most `tbuf_len` bytes from `src`, but if `strlen(tbuf) > 0` we could write over the end of `tbuf`.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+#define ENCODE_RAW(p_enc_conv, ch) \
+    (ch >= 0x20 && (p_enc_conv.vc_type == CONV_NONE \
+                    ? utf_printable(ch) \
+                    : ch < 0x7F))
```
`ch <= 0x7F`?

ZyX-I replied:
No. U+007F is DELETE which is not printable.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+    // Encode character as \u0000 if
+    // 1. It is an ASCII control character (0x0 .. 0x1F).
+    // 2. &encoding is not UTF-8 and code point is above 0x7F.
+    // 3. &encoding is UTF-8 and code point is printable according to
```
Missing `not`  in front of `printable`?

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+            xfree(tofree);
+            return FAIL;
+          } else if ((SURROGATE_HI_START <= ch && ch <= SURROGATE_HI_END)
+                     || (SURROGATE_LO_START <= ch && ch <= SURROGATE_LO_END)) {
```
Condition could be shortened to `SURROGATE_HI_START <= ch && ch <= SURROGATE_LO_END`

ZyX-I replied:
With magic numbers replaced by constants it is the job for the compiler.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+        } \
+      } \
+      if (conv_type == kMPConvDict) { \
+        vim_snprintf(ebuf, NUMBUFLEN + 6, "{...@%zu}", backref); \
```
Why not use `NUMBUFLEN + 7` for the buffer size?

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+        } \
+        default: { \
+          char numbuf[NUMBUFLEN]; \
+          vim_snprintf(numbuf, NUMBUFLEN - 1, "%g", flt_); \
```
Why `NUMBUFLEN - 1` instead of `NUMBUFLEN`?

ZyX-I committed:
eval/encode: Adjust buffer sizes passed to vim_snprintf

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+    }
+    if (line_end == end - 1) {
+      list_append_allocated_string(list, NULL);
+    }
```
This block could be moved out of the loop. 

We also could simplify the loop, if we continue the initial last list element before the loop. All other list elements do not need to be continued.

```diff
diff --git a/src/nvim/eval/encode.c b/src/nvim/eval/encode.c
index 0bde656..5f898ec 100644
--- a/src/nvim/eval/encode.c
+++ b/src/nvim/eval/encode.c
@@ -75,36 +75,38 @@ int encode_list_write(void *data, const char *buf, size_t len)
   const char *const end = buf + len;
   const char *line_end = buf;
   listitem_T *li = list->lv_last;
-  do {
+
+  // Continue the last list element
+  if (li != NULL) {
+    line_end = xmemscan(buf, NL, len);
+    if (line_end != buf) {
+      const size_t line_length = (size_t)(line_end - buf);
+      char *str = (char *)li->li_tv.vval.v_string;
+      const size_t li_len = (str == NULL ? 0 : strlen(str));
+      li->li_tv.vval.v_string = xrealloc(str, li_len + line_length + 1);
+      str = (char *)li->li_tv.vval.v_string + li_len;
+      memcpy(str, buf, line_length);
+      str[line_length] = 0;
+      memchrsub(str, NUL, NL, line_length);
+    }
+    line_end++;
+  }
+
+  while (line_end < end) {
     const char *line_start = line_end;
     line_end = xmemscan(line_start, NL, (size_t) (end - line_start));
     char *str = NULL;
     if (line_end != line_start) {
-      const size_t line_length = (size_t) (line_end - line_start);
-      if (li == NULL) {
-        str = xmemdupz(line_start, line_length);
-      } else {
-        const size_t li_len = (li->li_tv.vval.v_string == NULL
-                               ? 0
-                               : STRLEN(li->li_tv.vval.v_string));
-        li->li_tv.vval.v_string = xrealloc(li->li_tv.vval.v_string,
-                                           li_len + line_length + 1);
-        str = (char *) li->li_tv.vval.v_string + li_len;
-        memcpy(str, line_start, line_length);
-        str[line_length] = 0;
-      }
+      const size_t line_length = (size_t)(line_end - line_start);
+      str = xmemdupz(line_start, line_length);
       memchrsub(str, NUL, NL, line_length);
     }
-    if (li == NULL) {
-      list_append_allocated_string(list, str);
-    } else {
-      li = NULL;
-    }
-    if (line_end == end - 1) {
-      list_append_allocated_string(list, NULL);
-    }
+    list_append_allocated_string(list, str);
     line_end++;
-  } while (line_end < end);
+  }
+  if (line_end == end) {
+    list_append_allocated_string(list, NULL);
+  }
   return 0;
 }

```

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -0,0 +1,1291 @@
+#undef CONV_STRING
+#define CONV_STRING(buf, len) \
+    do { \
+      if (convert_to_json_string(gap, (const char *) (buf), (len)) != OK) { \
```
Some of the `CONV_STRING` macros are used with `STRLEN(buf)` as length argument (line 309). Can we be sure that `buf != NULL`? `convert_to_json_string()` has a check for `buf == NULL`.

ZyX-I committed:
eval/encode: Make sure that encoder can encode NULL variables

Adds two undocumented v: variables: _null_list and _null_dict because I do not 
know a reproducible way to get such lists (though I think I heard about this) 
and dictionaries (do not remember hearing about them). NULL strings are obtained 
using $XXX_UNEXISTENT_VAR_XXX.

Fixes crash in json_encode($XXX_UNEXISTENT_VAR_XXX). Other added tests worked 
fine before this commit.

ZyX-I committed:
eval/encode: Simplify loop in encode_list_write

Patch made up by oni-link.

Shougo commented:
I want to use this feature in my plugins.
Is it merge ready state?

ZyX-I commented:
I do not know any issues with this.

fatih commented:
I also need this for vim-go. Vim has it already and I have an upcoming feature based on it. If Neovim would merge this, I would like to release my upcoming change (guru support) both Vim and Neovim compatible.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -2841,9 +2841,8 @@ diff_hlID({lnum}, {col})				*diff_hlID()*
-		For a long |List| this is much faster than comparing the
-		length with zero.
+		items.  A Number is empty when its value is zero.  Special 
+		variable is empty when it is |v:false| or |v:none|.
```
@ZyX-I Since `v:none` was [removed](https://github.com/neovim/neovim/pull/4131/commits/b6c48798512f0ae2b50effc899da9ed2c8a5e029) it should not be mentioned here.

justinmk commented on src/nvim/message.h:
```diff
@@ -64,6 +64,18 @@
+/// First message
+extern MessageHistoryEntry *first_msg_hist;
+/// Last message
+extern MessageHistoryEntry *last_msg_hist;
```
why were these made `extern`? I don't see them being used outside of message.c.

justinmk commented on src/nvim/eval.c:
```diff
@@ -17037,18 +17037,6 @@ long get_vim_var_nr(int idx) FUNC_ATTR_PURE
-/// @warning if v: variable has reference counter it is not increased.
-///
-/// @param[in]  idx  Variable index, @see VimVarIndex.
-typval_T get_vim_var_tv(VimVarIndex idx) FUNC_ATTR_PURE
```
@ZyX-I Why remove this? The note about reference counter was a useful hint, which is more likely to be missed now. And this function reduced verbosity at the call site.

ZyX-I replied:
@justinmk This function was not ever used for anything, but special variables (v:true, v:false, v:null). So note about reference counter is not applicable anywhere where function was used. There is also no need to have such indirection, there are not too much places where function is needed.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -4897,9 +4897,13 @@ msgpackparse({list})				   {Nvim} *msgpackparse()*
-		boolean	One or zero. When dumping it is only checked that 
-			value is a |Number|.
+		nil	Zero, ignored when dumping.  This value cannot 
+			possibly apear in |msgpackparse()| output in Neovim 
```
typo

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -4897,9 +4897,13 @@ msgpackparse({list})				   {Nvim} *msgpackparse()*
+			possibly apear in |msgpackparse()| output in Neovim 
+			versions which have |v:null|.
+		boolean	One or zero.  When dumping it is only checked that 
+			value is a |Number|.  This value cannot possibly apear 
```
typo: apear

justinmk commented:
Let's merge it.

justinmk commented:
@ZyX-I Latest push has some failures like this:

```
...ovim/neovim/test/functional/eval/json_functions_spec.lua:312: Expected objects to be the same.
Passed in:
(table) {
 *[1] = userdata: 0x557ed1d0
  [2] = 1 }
Expected:
(table) {
  [2] = 1 }

stack traceback:
	...ovim/neovim/test/functional/eval/json_functions_spec.lua:312: in function <...ovim/neovim/test/functional/eval/json_functions_spec.lua:310>

[  FAILED  ] ...ovim/neovim/test/functional/eval/json_functions_spec.lua @ 310: json_decode() function parses containers (4.72 ms)
```

Also lint failures.

ZyX-I committed:
*: Fix preincrement lint errors

ZyX-I committed:
functests: Make json_functions_spec use new NIL where appropriate

justinmk commented:
@ZyX-I Just a heads up, still some failures.

justinmk commented:
Apparently CI cannot build when there are merge conflicts. Either this changed or I never noticed it before. Sorry.

ZyX-I commented:
Finally tests were fixed.
